On commence par charger les données dans notre fichier notebook

In [7]:
import sys
import os
os.chdir("..")
sys.path.append(os.getcwd())


In [8]:
from data.data_loader import load_data
df = load_data("C:/Users/akram/Processus_data/data/raw/synthetic_coffee_health_10000.csv")
df.head()

ID  Age  Gender  Country  Coffee_Intake  Caffeine_mg  Sleep_Hours  \
0   1   40    Male  Germany            3.5        328.1          7.5   
1   2   33    Male  Germany            1.0         94.1          6.2   
2   3   42    Male   Brazil            5.3        503.7          5.9   
3   4   53    Male  Germany            2.6        249.2          7.3   
4   5   32  Female    Spain            3.1        298.0          5.3   

  Sleep_Quality   BMI  Heart_Rate Stress_Level  Physical_Activity_Hours  \
0          Good  24.9          78          Low                     14.5   
1          Good  20.0          67          Low                     11.0   
2          Fair  22.7          59       Medium                     11.2   
3          Good  24.7          71          Low                      6.6   
4          Fair  24.1          76       Medium                      8.5   

  Health_Issues Occupation  Smoking  Alcohol_Consumption  
0           NaN      Other        0                    0  
1           NaN    Service        0                    0  
2          Mild     Office        0                    0  
3          Mild      Other        0                    0  
4          Mild    Student        0                    1

In [9]:
print(f"\nDimensions: {df.shape}")
print(f"\nType of Data:\n{df.dtypes}")


Dimensions: (10000, 16)

Type of Data:
ID                           int64
Age                          int64
Gender                      object
Country                     object
Coffee_Intake              float64
Caffeine_mg                float64
Sleep_Hours                float64
Sleep_Quality               object
BMI                        float64
Heart_Rate                   int64
Stress_Level                object
Physical_Activity_Hours    float64
Health_Issues               object
Occupation                  object
Smoking                      int64
Alcohol_Consumption          int64
dtype: object


In [10]:
print(f"\nMissing Values:\n{df.isnull().sum()}")


Missing Values:
ID                            0
Age                           0
Gender                        0
Country                       0
Coffee_Intake                 0
Caffeine_mg                   0
Sleep_Hours                   0
Sleep_Quality                 0
BMI                           0
Heart_Rate                    0
Stress_Level                  0
Physical_Activity_Hours       0
Health_Issues              5941
Occupation                    0
Smoking                       0
Alcohol_Consumption           0
dtype: int64


Comme vous avez remarquer qu'il existe un soucis , la colonne Heath_Issues contient 5941 valeur manquantes . On doit savoir pourquoi :)

In [15]:
Health_Issues_unique = df["Health_Issues"].unique()
print(Health_Issues_unique)

[nan 'Mild' 'Moderate' 'Severe']


Alors d'apres ce que j'ai compris c'est que la valeur "nan" signifie que la personne n'a pas de probleme de santer et non pas une cellule vide . 
Maintenant on commence par applique des statistique pour les variables numérique.

In [16]:
# Statistiques pour variables numériques spécifiques

num_vars = ['Age', 'Coffee_Intake', 'Caffeine_mg', 'Sleep_Hours', 
            'BMI', 'Heart_Rate', 'Physical_Activity_Hours', 'Smoking', 'Alcohol_Consumption']


for var in num_vars:
    print(f"\n{var}:")
    print(f"  Mean: {df[var].mean():.2f}")
    print(f"  Mediane: {df[var].median():.2f}")
    print(f"  Écart-type: {df[var].std():.2f}")
    print(f"  Min: {df[var].min():.2f} | Max: {df[var].max():.2f}")


Age:
  Mean: 34.95
  Mediane: 34.00
  Écart-type: 11.16
  Min: 18.00 | Max: 80.00

Coffee_Intake:
  Mean: 2.51
  Mediane: 2.50
  Écart-type: 1.45
  Min: 0.00 | Max: 8.20

Caffeine_mg:
  Mean: 238.41
  Mediane: 235.40
  Écart-type: 137.75
  Min: 0.00 | Max: 780.30

Sleep_Hours:
  Mean: 6.64
  Mediane: 6.60
  Écart-type: 1.22
  Min: 3.00 | Max: 10.00

BMI:
  Mean: 23.99
  Mediane: 24.00
  Écart-type: 3.91
  Min: 15.00 | Max: 38.20

Heart_Rate:
  Mean: 70.62
  Mediane: 71.00
  Écart-type: 9.82
  Min: 50.00 | Max: 109.00

Physical_Activity_Hours:
  Mean: 7.49
  Mediane: 7.50
  Écart-type: 4.32
  Min: 0.00 | Max: 15.00

Smoking:
  Mean: 0.20
  Mediane: 0.00
  Écart-type: 0.40
  Min: 0.00 | Max: 1.00

Alcohol_Consumption:
  Mean: 0.30
  Mediane: 0.00
  Écart-type: 0.46
  Min: 0.00 | Max: 1.00


Avant de commencer par appliquer des statistique sur des variables categorielles , il est conseil de les transformer sous forme
des valeurs numérique 

In [19]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder
colonne_non_numerique = ['Gender', 'Country', 'Sleep_Quality', 'Stress_Level',
 'Health_Issues', 'Occupation']

gender_unique = df[colonne_non_numerique[0]].unique()
country_unique = df[colonne_non_numerique[1]].unique()
sleep_quality_unique = df[colonne_non_numerique[2]].unique()
stress_level_unique = df[colonne_non_numerique[3]].unique()
Health_Issues_unique = df[colonne_non_numerique[4]].unique()
Occupation_unique = df[colonne_non_numerique[5]].unique()
print(gender_unique , country_unique , sleep_quality_unique , stress_level_unique , Health_Issues_unique , Occupation_unique)

label_encoders = {}  

for col in colonne_non_numerique:
    le = LabelEncoder()
    df[col] = df[col].fillna('Unknown')
    df[col] = le.fit_transform(df[col])
    label_encoders[col] = le

print(df.head())



['Male' 'Female' 'Other'] ['Germany' 'Brazil' 'Spain' 'Mexico' 'France' 'Canada' 'UK' 'Switzerland'
 'Netherlands' 'Italy' 'China' 'Japan' 'Belgium' 'Finland' 'Australia'
 'USA' 'Sweden' 'India' 'Norway' 'South Korea'] ['Good' 'Fair' 'Excellent' 'Poor'] ['Low' 'Medium' 'High'] [nan 'Mild' 'Moderate' 'Severe'] ['Other' 'Service' 'Office' 'Student' 'Healthcare']
   ID  Age  Gender  Country  Coffee_Intake  Caffeine_mg  Sleep_Hours  \
0   1   40       1        7            3.5        328.1          7.5   
1   2   33       1        7            1.0         94.1          6.2   
2   3   42       1        2            5.3        503.7          5.9   
3   4   53       1        7            2.6        249.2          7.3   
4   5   32       0       15            3.1        298.0          5.3   

   Sleep_Quality   BMI  Heart_Rate  Stress_Level  Physical_Activity_Hours  \
0              2  24.9          78             1                     14.5   
1              2  20.0          67             1  

Maintenant tout les variables sont numérique ! 

Traiter les doublons ! 

In [20]:
df.duplicated().sum()

np.int64(0)

Gerer les valeurs aberantes 

In [22]:
import pandas as pd

num_cols = ['Age', 'Coffee_Intake', 'Caffeine_mg', 'Sleep_Hours', 
            'BMI', 'Heart_Rate', 'Physical_Activity_Hours']

for col in num_cols:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    median = df[col].median()
    df[col] = df[col].apply(lambda x: median if x < lower_bound or x > upper_bound else x)
    
    
    print(f"{col}: {sum((df[col] < lower_bound) | (df[col] > upper_bound))} valeurs aberrantes corrigées")

for col in num_cols:
    print(f"{col}: Min={df[col].min()}, Max={df[col].max()}")



Age: 0 valeurs aberrantes corrigées
Coffee_Intake: 0 valeurs aberrantes corrigées
Caffeine_mg: 0 valeurs aberrantes corrigées
Sleep_Hours: 0 valeurs aberrantes corrigées
BMI: 0 valeurs aberrantes corrigées
Heart_Rate: 0 valeurs aberrantes corrigées
Physical_Activity_Hours: 0 valeurs aberrantes corrigées
Age: Min=18.0, Max=68.0
Coffee_Intake: Min=0.0, Max=6.5
Caffeine_mg: Min=0.0, Max=621.4
Sleep_Hours: Min=3.3, Max=10.0
BMI: Min=15.0, Max=34.5
Heart_Rate: Min=50.0, Max=96.0
Physical_Activity_Hours: Min=0.0, Max=15.0


Heureusement il existe pas des valeurs aberrantes , donc nos données sont totalement fiable 

Il est recommander de standariser les variable numérique , pour cela on va utiliser la méthode StandardScaler

In [23]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
df[num_cols] = scaler.fit_transform(df[num_cols])
print(df.head())

   ID       Age  Gender  Country  Coffee_Intake  Caffeine_mg  Sleep_Hours  \
0   1  0.467675       1        7       0.709187     0.676561     0.707250   
1   2 -0.168481       1        7      -1.048907    -1.055965    -0.368604   
2   3  0.649434       1        2       1.975015     1.976697    -0.616878   
3   4  1.649108       1        7       0.076273     0.092389     0.541734   
4   5 -0.259361       0       15       0.427892     0.453702    -1.113426   

   Sleep_Quality       BMI  Heart_Rate  Stress_Level  Physical_Activity_Hours  \
0              2  0.250595    0.783058             1                 1.625265   
1              2 -1.027844   -0.361824             1                 0.814134   
2              1 -0.323398   -1.194466             2                 0.860485   
3              2  0.198414    0.054496             1                -0.205573   
4              1  0.041870    0.574897             2                 0.234755   

   Health_Issues  Occupation  Smoking  Alcohol_Con